File: WordBooksV2/app/src/main/java/com/callor/word/persistance/WordRepository.java
package com.callor.word.persistance;

import android.app.Application;
import androidx.lifecycle.LiveData;
import com.callor.word.model.WordVO;
import java.util.List;

public class WordRepository {

    private WordDao wordDao;
    private LiveData<List<WordVO>> wordList;

    public WordRepository(Application application) {
        WordDatabase db = WordDatabase.getInstance(application);
        wordDao = db.wordDao();
        wordList = wordDao.selectAll();
    }

    public LiveData<List<WordVO>> selectAll() {
        return wordDao.selectAll();
    }

    public void insert(WordVO wordVO) {
        /*
         insert SQL 은 실제로 시간이 많이 소요되고
         오류가 발생할 확률이 높기 때문에
         수행할때 background 에서 수행하라

         JS 에서 함수의 매개변수로 함수를 전달하는 코드
         onClick={ ()=>onClickHandler() }
         onClick={ onClickHandler }

         java 1.8 이상에서는 자바의 method 를 다른 method 에
         전달할 수 있다

         execute() method 에게 wordDao.insert(데이터) method 를 보내서
         execute() 가 wordDao.insert() method 를 실행하도록 지시하기

         ()->wordDao.insert(wordVO) : lambda 코드로 method 변환시켜서
         execute() method 에게 전달하고 대신 실행하도록 지시하기

         */
        WordDatabase.databaseExcutor.execute(
                ()-> wordDao.insert(wordVO)
        );
    }


}

File: WordBook/app/src/main/java/com/callor/word/persistance/WordRepository.kt
package com.callor.word.persistance

import androidx.annotation.WorkerThread
import com.callor.word.model.Word
import kotlinx.coroutines.flow.Flow
/*
WordDao 를 생성자로 주입받기

Room 은 별도의 thread(WorkThread)에서 모든 쿼리를 실행한다
suspend 는 coroutines 나 다른 정지 함수에서 호출하도록 한다는 선언 

 */
class WordRepository(private val wordDao: WordDao) {

    // List<Word> 의 데이터가 변경되면
    // 화면에 보여지는 View 에게 알림 전파하기
    val allWords : Flow<List<Word>> = wordDao.selectAll()

    /*
    insert 작업을 수행할때 main thread 에서 실행하지 않도록 설정
    insert 작업의 경우 생각보다 많은 시간이 소요된다
    이러한 작업을 main thread 에서 수행하면
    화면이 멈춘 것처럼 보일수 있기 때문에
    background 에서 수행하도록 설정
     */
    @Suppress("RedundantSuspendModifier")
    @WorkerThread
    suspend fun insert(word:Word) {
        wordDao.insert(word)
    }


}